### ISSUE TITLE
关于训练集的face mask提取

### ISSUE THREAD

**Issue by potentialming** [Francis-Rings/StableAnimator#46]
尝试了你们的推理代码，效果真的太棒了。我在尝试训练它，但是在使用你们的提供的face_mask_extraction.py批处理数据集以获取faces时，发现速度太慢了。你们是否使用了速度更快的方式创建faces？

**Comment by potentialming** [comment]
或者可以分享下你们获取训练集的一些细节吗？不胜感激！

**Comment by Francis-Rings** [comment]
Hi, please verify whether the face detector is running on the GPU. It becomes significantly slower when the detector runs on the CPU instead of the GPU.

**Comment by Francis-Rings** [comment]
> 或者可以分享下你们获取训练集的一些细节吗？不胜感激！

You can collect your own training videos from the social media platforms, such as Bilibili, TikTok, and Youtube.

**Comment by potentialming** [comment]
我确认我使用了GPU，我使用了公开的TiTok数据集，大概300多个视频，每个视频大概300帧左右，在一张24G 3090上预计要处理30多个小时甚至更多。请问这正常吗？我的处理代码大概是这样。
`import numpy as np
import torch
from facexlib.parsing import init_parsing_model
from facexlib.utils.face_restoration_helper import FaceRestoreHelper
from insightface.app import FaceAnalysis
import cv2
import os
from concurrent.futures import ThreadPoolExecutor

def get_face_masks(image_path, save_path, app, face_helper, height=904, width=512):
    # 如果目标文件已经存在，直接返回
    if os.path.exists(save_path):
        print(f"Face mask already exists for {image_path}. Skipping...")
        return
    
    # 读取图像
    image_1 = cv2.imread(image_path)
    height, width = image_1.shape[:2]
    image_bgr_1 = cv2.cvtColor(image_1, cv2.COLOR_RGB2BGR)
    image_info_1 = app.get(image_bgr_1)

    # 初始化掩膜
    mask_1 = np.zeros((height, width), dtype=np.uint8)

    # 使用 FaceAnalysis 提取面部区域
    if len(image_info_1) > 0:
        print(f"FaceAnalysis detected face in {image_path}")
        for info in image_info_1:
            x_1 = info['bbox'][0]
            y_1 = info['bbox'][1]
            x_2 = info['bbox'][2]
            y_2 = info['bbox'][3]
            cv2.rectangle(mask_1, (int(x_1), int(y_1)), (int(x_2), int(y_2)), (255), thickness=cv2.FILLED)
        cv2.imwrite(save_path, mask_1)
    else:
        # 如果 FaceAnalysis 未检测到面部，使用 FaceRestoreHelper
        face_helper.clean_all()
        with torch.no_grad():
            bboxes = face_helper.face_det.detect_faces(image_bgr_1, 0.97)
        if len(bboxes) > 0:
            print(f"FaceRestoreHelper detected face in {image_path}")
            for bbox in bboxes:
                cv2.rectangle(mask_1, (int(bbox[0]), int(bbox[1])), (int(bbox[2]), int(bbox[3])), (255), thickness=cv2.FILLED)
            cv2.imwrite(save_path, mask_1)
        else:
            print(f"No face detected in {image_path}")
            mask_1[:] = 255  # 如果未检测到人脸，填充白色
            cv2.imwrite(save_path, mask_1)

def process_images_in_folder(folder_path, app, face_helper):
    # 创建一个线程池来并行处理图像
    with ThreadPoolExecutor(max_workers=13) as executor:  # 根据您的机器性能调整 max_workers
        # 遍历 vec 文件夹中的每个子文件夹
        for subfolder_name in os.listdir(folder_path):
            subfolder_path = os.path.join(folder_path, subfolder_name)

            # 判断是否为文件夹
            if os.path.isdir(subfolder_path):
                # 确保 images 文件夹存在
                images_folder = os.path.join(subfolder_path, 'images')
                if os.path.exists(images_folder):
                    face_subfolder_path = os.path.join(subfolder_path, 'faces')
                    if not os.path.exists(face_subfolder_path):
                        os.makedirs(face_subfolder_path)
                        print(f"Folder created: {face_subfolder_path}")
                    else:
                        print(f"Folder already exists: {face_subfolder_path}")
                    
                    # 遍历 images 文件夹中的所有 PNG 图像文件
                    for file in os.listdir(images_folder):
                        if file.endswith('.png'):
                            file_path = os.path.join(images_folder, file)
                            face_save_path = os.path.join(face_subfolder_path, file)  # 保持相同文件名
                            executor.submit(get_face_masks, file_path, face_save_path, app, face_helper)
                            print(f"Submitted face extraction task for: {file_path}")

def main():
    # 直接在这里定义文件夹路径
    folder_path = '/root/autodl-tmp/eTikTok/vec'  # 这里替换为您实际的路径

    # 初始化 FaceAnalysis 和 FaceRestoreHelper
    app = FaceAnalysis(
        name='antelopev2', root='.', providers=['CUDAExecutionProvider', 'CPUExecutionProvider']
    )
    app.prepare(ctx_id=0, det_size=(640, 640))
    
    face_helper = FaceRestoreHelper(
        upscale_factor=1,
        face_size=512,
        crop_ratio=(1, 1),
        det_model='retinaface_resnet50',
        save_ext='png',
        device="cuda",
    )
    face_helper.face_parse = init_parsing_model(model_name='bisenet', device="cuda")

    print(f"Processing vec folder: {folder_path}")
    
    # 处理文件夹中的所有子文件夹
    process_images_in_folder(folder_path, app, face_helper)

if __name__ == "__main__":
    main()
`
能帮我看看原因吗？

**Comment by Francis-Rings** [comment]
Since face mask extraction does not require a large amount of GPU memory, you can run multiple instances of face_mask_extraction.py in parallel (using `nohup`) to improve efficiency.

**Comment by potentialming** [comment]
目前我的代码是使用的多线程，不知道并行运行多个 face_mask_extraction.py会不会更好，我会尝试一下，感谢解惑。
